File: ext/java_bin/ext/parser.c
#include "parser.h"

/*
 * variables
 */
static VALUE rb_mJavaBin;
static VALUE rb_mExt;
static VALUE rb_cParser;

static ID    i_At; // Time#at用
static ID    i_Utc; // Time#utc用
#ifdef HAVE_RUBY_ENCODING_H
  static rb_encoding* rb_encUtf8;
#endif

static VALUE jbp_numFound;
static VALUE jbp_start;
static VALUE jbp_maxScore;
static VALUE jbp_docs;

static VALUE jbp_terminator;

/*
 * javabinフォーマット読み込み関数群
 */
static inline void JavaBinParser_extend_cache(JAVA_BIN_PARSER* ptr) {
  int next_size;

  next_size = ((ptr->cache == NULL) ? 64 : ptr->cache_size * 2);
  ptr->cache = (VALUE*) realloc(ptr->cache, next_size * sizeof(VALUE));
  if (!ptr->cache) {
    rb_raise(rb_eRuntimeError, "JavaBinParser_extend_cache - allocate error");
  }
  ptr->cache_size = next_size;
}

static inline int32_t JavaBinParser_read_v_int(JAVA_BIN_PARSER* ptr) {
  unsigned char byte;
  int32_t result;
  int shift;
  byte = _getbyte(ptr);
  result = byte & 0x7f;
  for (shift = 7; (byte & 0x80) != 0; shift += 7) {
    byte = _getbyte(ptr);
    result |= (((int32_t)(byte & 0x7f)) << shift);
  }
  return result;
}

static inline int64_t JavaBinParser_read_v_long(JAVA_BIN_PARSER* ptr) {
  unsigned char byte;
  int64_t result;
  int shift;
  byte = _getbyte(ptr);
  result = byte & 0x7f;
  for (shift = 7; (byte & 0x80) != 0; shift += 7) {
    byte = _getbyte(ptr);
    result |= (((int64_t)(byte & 0x7f)) << shift);
  }
  return result;
}

static inline int JavaBinParser_read_size(JAVA_BIN_PARSER* ptr) {
  int size;
  size = (ptr->tag_byte & 0x1f);
  if (size == 0x1f) {
    size += JavaBinParser_read_v_int(ptr);
  }
  return size;
}

static inline VALUE JavaBinParser_read_small_int(JAVA_BIN_PARSER* ptr) {
  int32_t result;
  result = ptr->tag_byte & 0x0f;
  if ((ptr->tag_byte & 0x10) != 0) {
    result = ((JavaBinParser_read_v_int(ptr) << 4) | result);
  }
  return INT2NUM(result);
}

static inline VALUE JavaBinParser_read_small_long(JAVA_BIN_PARSER* ptr) {
  int64_t result;
  result = ptr->tag_byte & 0x0f;
  if ((ptr->tag_byte & 0x10) != 0) {
    result = ((JavaBinParser_read_v_long(ptr) << 4) | result);
  }
  return LL2NUM(result);
}

static inline VALUE JavaBinParser_read_string(JAVA_BIN_PARSER* ptr) {
  int size, i;
  int start;
  unsigned char b;
  size = JavaBinParser_read_size(ptr);
  start = ptr->current;
  for (i = 0; i < size; i++) {
    /* HINT. read utf-8 char */
    b = _getbyte(ptr);
    if ((b & 0x80) == 0) {
    } else if ((b & 0xE0) == 0xC0) {
      _skipbytes(ptr, 1);
    } else {
      _skipbytes(ptr, 2);
    } /* TODO 4byte以上のケース? */
  }
  return _utf8_string((const char*) &ptr->data[start], ptr->current - start); 
}

static inline VALUE JavaBinParser_read_array(JAVA_BIN_PARSER* ptr) {
  int size, i;
  VALUE array;
  size  = JavaBinParser_read_size(ptr);
  array = rb_ary_new();
  for (i = 0; i < size; i++) {
    rb_ary_push(array,JavaBinParser_read_val(ptr));
  }
  return array;
}

static inline VALUE JavaBinParser_read_extern_string(JAVA_BIN_PARSER* ptr) {
  int size;
  VALUE value;
  size = JavaBinParser_read_size(ptr);
  if (size == 0) {
    /* rubyの文字列 */
    value = JavaBinParser_read_val(ptr);
    OBJ_FREEZE(value);
    /* 参照文字列として文字列を保持 */
    ptr->cache[ptr->cache_index++] = value;
    /* 必要があれば参照文字列用のcacheを拡張する */
    if (ptr->cache_size <= ptr->cache_index) {
      JavaBinParser_extend_cache(ptr);
    }
    return value;
  } else {
    //return rb_str_new4(ptr->cache[size - 1]);   // freeze共有
    //return rb_str_new3(ptr->cache[size - 1]); // 共有(変更があったら分裂)
    return ptr->cache[size - 1];              // 同じ物
    //return rb_str_dup(ptr->cache[size - 1]);  // コピー
  }
}

static inline VALUE JavaBinParser_read_ordered_map(JAVA_BIN_PARSER* ptr) {
  int size, i;
  VALUE hash, key, value;
  size = JavaBinParser_read_size(ptr);
  hash = rb_hash_new();
  for (i = 0; i < size; i++) {
    key   = JavaBinParser_read_val(ptr);
    value = JavaBinParser_read_val(ptr);
    rb_hash_aset(hash, key, value);
  }
  return hash;
}

static inline VALUE JavaBinParser_read_byte(JAVA_BIN_PARSER* ptr) {
  int8_t c;
  _readnumeric(ptr, c);
  return INT2FIX(*((int8_t*)&c));
}

static inline VALUE JavaBinParser_read_short(JAVA_BIN_PARSER* ptr) {
  u_int16_t c;
  _readnumeric(ptr, c);
  c = _swap_16(c);
  return INT2FIX(*((int16_t*)&c));
}

static inline VALUE JavaBinParser_read_int(JAVA_BIN_PARSER* ptr) {
  u_int32_t c;
  _readnumeric(ptr, c);
  c = _swap_32(c);
  return INT2NUM(*((int32_t*)&c));
}

static inline VALUE JavaBinParser_read_long(JAVA_BIN_PARSER* ptr) {
  u_int64_t c;
  _readnumeric(ptr, c);
  c = _swap_64(c);
  return LL2NUM(*((int64_t*)&c));
}

static inline VALUE JavaBinParser_read_date(JAVA_BIN_PARSER* ptr) {
  u_int64_t c;
  _readnumeric(ptr, c);
  c = _swap_64(c);
  return rb_funcall(rb_funcall(rb_cTime, i_At, 1, ULL2NUM(c / 1000)), i_Utc, 0);
}

static inline VALUE JavaBinParser_read_float(JAVA_BIN_PARSER* ptr) {
  JAVA_BIN_FLOAT c;
  _readnumeric(ptr, c.i_value);
  c.i_value = _swap_32(c.i_value);
  return rb_float_new((double)c.f_value);
}

static inline VALUE JavaBinParser_read_double(JAVA_BIN_PARSER* ptr) {
  JAVA_BIN_DOUBLE c;
  _readnumeric(ptr, c.i_value);
  c.i_value = _swap_64(c.i_value);
  return rb_float_new(c.d_value);
}

static inline VALUE JavaBinParser_read_byte_array(JAVA_BIN_PARSER* ptr) {
  int size, i;
  VALUE array;
  size = JavaBinParser_read_v_int(ptr);
  array = rb_ary_new();
  for (i = 0; i < size; i++) {
    rb_ary_push(array, INT2FIX(_getbyte(ptr)));
  }
  return array;
}

static inline VALUE JavaBinParser_read_map(JAVA_BIN_PARSER* ptr) {
  int size, i;
  VALUE hash, key, value;
  size = JavaBinParser_read_v_int(ptr);
  hash = rb_hash_new();
  for (i = 0; i < size; i++) {
    key   = JavaBinParser_read_val(ptr);
    value = JavaBinParser_read_val(ptr);
    rb_hash_aset(hash, key, value);
  }
  return hash;
}

static inline VALUE JavaBinParser_read_iterator(JAVA_BIN_PARSER* ptr) {
  VALUE array, value;
  array = rb_ary_new();
  while (1) {
    value = JavaBinParser_read_val(ptr);
    if (value == END_OBJ) {
      break;
    }
    rb_ary_push(array, value);
  }
  return array;
}

static inline VALUE JavaBinParser_read_solr_doc_list(JAVA_BIN_PARSER* ptr) {
  VALUE hash, array;
  hash = rb_hash_new();
  array = JavaBinParser_read_val(ptr);
  rb_hash_aset(hash, jbp_numFound, rb_ary_entry(array, 0));
  rb_hash_aset(hash, jbp_start,    rb_ary_entry(array, 1));
  rb_hash_aset(hash, jbp_maxScore, rb_ary_entry(array, 2));
  rb_hash_aset(hash, jbp_docs,     JavaBinParser_read_val(ptr));
  return hash;
}

static VALUE JavaBinParser_read_val(JAVA_BIN_PARSER* ptr) {
  ptr->tag_byte = _getbyte(ptr);
  switch (ptr->tag_byte >> 5) { /* unsignedなので論理シフト */
    case SHIFTED_STR:
      return JavaBinParser_read_string(ptr);
    case SHIFTED_ARR:
      return JavaBinParser_read_array(ptr);
    case SHIFTED_EXTERN_STRING:
      return JavaBinParser_read_extern_string(ptr);
    case SHIFTED_ORDERED_MAP:
    case SHIFTED_NAMED_LST:
      return JavaBinParser_read_ordered_map(ptr);
    case SHIFTED_SINT:
      return JavaBinParser_read_small_int(ptr);
    case SHIFTED_SLONG:
      return JavaBinParser_read_small_long(ptr);
  }

  switch(ptr->tag_byte) {
    case BYTE:
      return JavaBinParser_read_byte(ptr);
    case SHORT:
      return JavaBinParser_read_short(ptr);
    case DOUBLE:
      return JavaBinParser_read_double(ptr);
    case INT:
      return JavaBinParser_read_int(ptr);
    case LONG:
      return JavaBinParser_read_long(ptr);
    case FLOAT:
      return JavaBinParser_read_float(ptr);
    case DATE:
      return JavaBinParser_read_date(ptr);
    case BYTEARR:
      return JavaBinParser_read_byte_array(ptr);
    case NULL_MARK:
      return Qnil;
    case BOOL_TRUE:
      return Qtrue;
    case BOOL_FALSE:
      return Qfalse;
    case MAP:
      return JavaBinParser_read_map(ptr);
    case ITERATOR:
      return JavaBinParser_read_iterator(ptr);
    case END:
      return END_OBJ; 
    case SOLRDOC:
      return JavaBinParser_read_val(ptr);
    case SOLRDOCLST:
      return JavaBinParser_read_solr_doc_list(ptr);
    default:
      rb_raise(rb_eRuntimeError, "JavaBinParser_read_val - unknown tag type");
  }
}

/*
 * rubyメソッド
 */
static VALUE rb_cParser_parse(VALUE self, VALUE data) {
  JAVA_BIN_PARSER* ptr;
  char* ptrData;
  int   dataLen;

  Data_Get_Struct(self, JAVA_BIN_PARSER, ptr);

  /* 引数処理 */
  if (TYPE(data) != T_STRING) {
    rb_raise(rb_eRuntimeError, "rb_cParser_parse - data is not String.");
  }

  SafeStringValue(data);
  ptrData = RSTRING_PTR(data);
  dataLen = RSTRING_LEN(data);

  /* 引数チェック */
  if (ptrData == NULL || dataLen == 0) {
    rb_raise(rb_eRuntimeError, "rb_cParser_parse - data is empty.");
  }

  ptr->data = (unsigned char*)ptrData;
  ptr->data_len = dataLen;

  /* version check */
  if (ptr->data[0] != 0x01) {
    rb_raise(rb_eRuntimeError, "rb_cParser_parse - not supported version [%d]", ptr->data[0]);
  }

  ptr->current  = 1;   /* VERSIONをとばした */
  ptr->tag_byte = 0;

  /*
   * 参照文字列保持、既に確保している場合は使いまわす。
   * 確保していない場合は初回分を確保
   */
  ptr->cache_index = 0;
  if (!ptr->cache) {
    JavaBinParser_extend_cache(ptr);
  }
 
  return JavaBinParser_read_val(ptr);
}

static VALUE rb_cParser_initialize(VALUE self) {
  JAVA_BIN_PARSER* ptr;

  /* データの初期化 */
  ptr = (JAVA_BIN_PARSER*) malloc(sizeof(JAVA_BIN_PARSER));
  if (!ptr) {
    rb_raise(rb_eRuntimeError, "rb_cParser_initialize - allocate error");
  }
  DATA_PTR(self) = ptr;

  /* 参照文字列の準備 */
  ptr->cache = NULL;
  ptr->cache_size = 0;
  ptr->cache_index = 0;
 
  return self;
}

static void JavaBinParser_free(JAVA_BIN_PARSER* ptr) {
  if (ptr) {
    if (ptr->cache) {
      free(ptr->cache);
    }
    free(ptr);
  }
}

static void JavaBinParser_mark(JAVA_BIN_PARSER* ptr) {
  int i;
  if (ptr) {
    for (i = 0; i < ptr->cache_index; i++) {
      rb_gc_mark_maybe(ptr->cache[i]);
    }
  }
}

static VALUE JavaBinParser_alloc(VALUE klass) {
  return Data_Wrap_Struct(klass, JavaBinParser_mark, JavaBinParser_free, NULL);
}

/*
 * エントリーポイント
 */
void Init_parser(void) {
  i_At = rb_intern("at");
  i_Utc = rb_intern("utc");
#ifdef HAVE_RUBY_ENCODING_H
  rb_encUtf8 = rb_utf8_encoding();
#endif

  /* クラス定義 */
  rb_mJavaBin = rb_define_module("JavaBin");
  rb_mExt     = rb_define_module_under(rb_mJavaBin, "Ext");
  rb_cParser  = rb_define_class_under(rb_mExt, "Parser", rb_cObject);
 
  /* メモリーアロケーター設定 */
  rb_define_alloc_func(rb_cParser, JavaBinParser_alloc);
  /* コンストラクタ */
  rb_define_method(rb_cParser, "initialize", rb_cParser_initialize, 0);
  /* parseメソッド*/
  rb_define_method(rb_cParser, "parse", rb_cParser_parse, 1);

  /* 繰り返し使うので変数&定数化 */
  /* HINT 定数にするとGCの対象にならないっぽいので */
  rb_define_const(rb_cParser, "NUM_FOUND", rb_str_new2("numFound"));
  jbp_numFound = rb_const_get(rb_cParser, rb_intern("NUM_FOUND"));
  rb_define_const(rb_cParser, "START", rb_str_new2("start"));
  jbp_start    = rb_const_get(rb_cParser, rb_intern("START")); 
  rb_define_const(rb_cParser, "MAX_SCORE", rb_str_new2("maxScore"));
  jbp_maxScore = rb_const_get(rb_cParser, rb_intern("MAX_SCORE"));
  rb_define_const(rb_cParser, "DOCS", rb_str_new2("docs"));
  jbp_docs     = rb_const_get(rb_cParser, rb_intern("DOCS"));
  OBJ_FREEZE(jbp_numFound);
  OBJ_FREEZE(jbp_start);
  OBJ_FREEZE(jbp_maxScore);
  OBJ_FREEZE(jbp_docs);
 
  rb_define_const(rb_cParser, "TERMINATOR", rb_str_new2("terminator")); // iterator終端用のオブジェクト
  jbp_terminator = rb_const_get(rb_cParser, rb_intern("TERMINATOR"));
  OBJ_FREEZE(jbp_terminator);
}

File: lib/java_bin/pure/parser.rb
# vim:fileencoding=utf-8

module JavaBin
  module Pure
  NULL       = 0
  BOOL_TRUE  = 1
  BOOL_FALSE = 2
  BYTE       = 3
  SHORT      = 4
  DOUBLE     = 5
  INT        = 6
  LONG       = 7
  FLOAT      = 8
  DATE       = 9
  MAP        = 10
  SOLRDOC    = 11
  SOLRDOCLST = 12
  BYTEARR    = 13
  ITERATOR   = 14
  TERM       = 15 #END = 15

  #TAG_AND_LEN  = (1 << 5)
  STR           = (1 << 5)
  SINT          = (2 << 5)
  SLONG         = (3 << 5)
  ARR           = (4 << 5)
  ORDERED_MAP   = (5 << 5)
  NAMED_LST     = (6 << 5)
  EXTERN_STRING = (7 << 5)

  # TODO 論理シフト
  SHIFTED_STR           = STR >> 5
  SHIFTED_ARR           = ARR >> 5
  SHIFTED_EXTERN_STRING = EXTERN_STRING >> 5
  SHIFTED_ORDERED_MAP   = ORDERED_MAP >> 5
  SHIFTED_NAMED_LST     = NAMED_LST >> 5
  SHIFTED_SINT          = SINT >> 5
  SHIFTED_SLONG         = SLONG >> 5

    class Perser

  VERSION  = 1
  TERM_OBJ = :term_obj

  def self.unmarshal(input)
    array = input.bytes.to_a
    check_version(array[0])
    Reader.new(array).process
  end

  def self.check_version(byte)
    return true if VERSION == byte
    raise "unsupported version #{byte}"
  end

  class Reader
 
    #attr_reader :tag_byte, :input, :current

    def initialize(input)
      @input = input
      @current = 1 # HINT VERSIONをとばす
      @tag_byte = nil
    end

    def process
      read_val
    end

    private
    def getbyte
      ret = @input[@current]
      @current += 1
      ret
    end
    
    def getbytes(size)
      ret = @input[@current...(@current + size)]
      @current += size
      ret
    end
 
    def read_val
      @tag_byte = getbyte
      case (@tag_byte >> 5) # TODO 論理シフト
        when SHIFTED_STR
          return read_chars
        when SHIFTED_ARR
          size = read_size
          array = Array.new(size)
          size.times { |i| array[i] = read_val }
          return array
        when SHIFTED_EXTERN_STRING
          size = read_size
          if size == 0
            str = read_val
            @exts ||= []
            @exts << str
            return str
          else
            return @exts[size - 1].dup
          end
        when SHIFTED_ORDERED_MAP, SHIFTED_NAMED_LST
          size = read_size
          hash = {}
          size.times do
            k = read_val
            v = read_val
            hash[k] = v
          end
          return hash
        when SHIFTED_SINT
          return read_small_int
        when SHIFTED_SLONG
          return read_small_int
      end
  
      case @tag_byte
        when NULL
          return nil
        when BOOL_TRUE
          return true
        when BOOL_FALSE
          return false
        when BYTE
          return getbytes(1).pack("C*").unpack("c")[0]
        when SHORT
          return getbytes(2).reverse.pack("C*").unpack("s")[0]
        when DOUBLE
          return getbytes(8).pack("C*").unpack("G")[0]
        when INT
          return getbytes(4).reverse.pack("C*").unpack("i")[0]
        when LONG
          return getbytes(8).reverse.pack("C*").unpack("q")[0]
        when FLOAT
          return getbytes(4).pack("C*").unpack("g")[0]
        when DATE
          x = getbytes(8).reverse.pack("C*").unpack("q")[0]
          return Time.at(x/1000)
        when MAP
          size = read_v_int
          hash = {}
          size.times do
            k = read_val
            v = read_val
            hash[k] = v
          end
          return hash
        when BYTEARR
          size = read_v_int
          return getbytes(size)
        when ITERATOR
          array = []
          while true
            i = read_val
            break if i == TERM_OBJ
            array << i
          end
          return array
        when TERM
          return TERM_OBJ
        when SOLRDOC
          return read_val
        when SOLRDOCLST
          result = {}
          list = read_val
          result['numFound'] = list[0]
          result['start']    = list[1]
          result['maxScore'] = list[2]
          result['docs'] = read_val
          return result
      end
    end
  
    def read_v_int
      byte = getbyte
      result = byte & 0x7f
      shift = 7
      while (byte & 0x80) != 0
        byte = getbyte
        result |= ((byte & 0x7f) << shift)
        shift += 7
      end
      result 
    end
  
    def read_size
      size = (@tag_byte & 0x1f)
      size += read_v_int if (size == 0x1f)
      size
    end
  
    def read_small_int
      result = @tag_byte & 0x0F
      result = ((read_v_int << 4) | result) if ((@tag_byte & 0x10) != 0)
      result
    end
  
    def read_chars
      size = read_size
      str = ''
      size.times {
        # HINT. read utf-8 char
        b = getbyte
        if ((b & 0x80) == 0)
          str << b
        elsif ((b & 0xE0) != 0xE0)
          str << (((b & 0x1F) << 6) | (getbyte & 0x3F))
        else
          str << (((b & 0x0F) << 12) | ((getbyte & 0x3F) << 6) | (getbyte & 0x3F))
        end
      }
      str.force_encoding('utf-8') if str.respond_to? :force_encoding
      str
    end

  end

end

